### PR TITLE
[3152] Add missing fields banner to registered trainees

### PR DESCRIPTION
**Pull request by AbigailMcP**
### Context

https://trello.com/c/905rEvfQ/3152-l-add-outstanding-actions-to-trainees-missing-fields-banner

### Changes proposed in this pull request

- Adds a generic 'dumb' `NoticeBanner` component which renders and heading and some content with the correct stying, if content is present.
- Adds a mechanism to `Submissions::BaseValidator` by which extra validators can be added to its subclasses. This is needed because we now have a form `TraineeStartDateForm` which needs to be validated for determining whether a trainee has missing fields (and eventually for QTS recommendation) BUT does not need to be validated for TRN in the `TrnValidator`.
- Adds the extra validator to `MissingDataValidator`.
- Adds a new `MissingDataBannerView` which transforms the missing fields for a trainee into a bulleted list of links, linking to the page containing the form that needs to be completed.

**For free!**
- The addition of the extra validator to `MissingDataValidator` should mean that the 'INCOMPLETE' tag now correctly renders for registered trainees who's **course has already started** but **missing a trainee start date**. Up until now they would have looked complete.

Screenshot 2021-11-18 at 14 23 22

### Out of scope
- The grouping of missing field links into one link if a whole form is incomplete. I will card this up separately.
- The trainee status inset text - this is already carded up separately https://trello.com/c/10ILGPQN/3241-m-add-outstanding-actions-to-trainees-trainee-status-inset-text

### Guidance to review

**These trainees should have the correct banner content, plus an incomplete label:**

For a trainee with no start date, ITT in the past:
https://register-pr-1745.london.cloudapps.digital/trainees/regs7T1w7o7Eh9C3vyVcayPV

For a trainee missing a training initiative:
https://register-pr-1745.london.cloudapps.digital/trainees/pLMycBxBZ3tGuwK7oJTFSf6W

For a trainee missing an initiative and funding:
https://register-pr-1745.london.cloudapps.digital/trainees/hSQ1Qa4cXNehKcrXGXTyfC5n

**This trainee should have no banner and no incomplete tag (ITT in the future)**
https://register-pr-1745.london.cloudapps.digital/trainees/y1fbzXYmEAMJS1gNfk9L6ymF

### CODE REVIEW

**Comment by AbigailMcP** [comment]
[[M] Add 'Outstanding actions' to trainees - trainee status inset text](https://trello.com/c/10ILGPQN/3241-m-add-outstanding-actions-to-trainees-trainee-status-inset-text)

File: app/components/notice_banner/view.rb [hunk 1]
```diff
@@ -0,0 +1,13 @@
+# frozen_string_literal: true
+
+module NoticeBanner
+  class View < GovukComponent::Base
+    renders_one :header
+
+    def initialize; end
+
+    def render?
+      content.present?
```
Comment (darokel) [inline app/components/notice_banner/view.rb:10]: 🥇

File: app/forms/submissions/missing_data_validator.rb [hunk 2]
```diff
@@ -2,16 +2,30 @@
...
+    missing_data_validator :trainee_start_date, form: "TraineeStartDateForm", if: :course_already_started?
+
     def missing_fields
       forms.map(&:missing_fields).flatten.uniq
     end
 
-    def forms
-      validator_keys.map { |key| validator(key) }
+    def course_already_started?
+      !trainee.course_start_date&.future?
```
Comment (darokel) [inline app/forms/submissions/missing_data_validator.rb:20]:
```suggestion
!trainee.starts_course_in_the_future?
```

https://github.com/DFE-Digital/register-trainee-teachers/blob/main/app/models/trainee.rb#L321

File: spec/view_objects/missing_data_banner_view_spec.rb [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    context "when there is no missing data" do
+      let(:trainee) { create(:trainee, :completed) }
+      let(:missing_fields) { Submissions::MissingDataValidator.new(trainee: trainee).missing_fields }
+
+      it "returns nothing" do
+        expect(subject).to be_falsey
+      end
+    end
+
+    context "when the trainee is missing data -" do
```
Comment (darokel) [inline spec/view_objects/missing_data_banner_view_spec.rb:20]:
```suggestion
context "when the trainee is missing data" do
```

File: spec/view_objects/missing_data_banner_view_spec.rb [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+      let(:trainee) { create(:trainee, :completed, :course_start_date_in_the_past, field => nil) }
+      let(:missing_fields) { Submissions::MissingDataValidator.new(trainee: trainee).missing_fields }
```
Comment (darokel) [inline spec/view_objects/missing_data_banner_view_spec.rb:22]: Could this be moved further up for both contexts if it's the same?

**Review by darokel — APPROVED** [review]
Left a couple of tiny comments but the code looks good 🚀

**Comment by AbigailMcP** [comment]
[[L] Add 'outstanding actions' to trainees - missing fields banner](https://trello.com/c/905rEvfQ/3152-l-add-outstanding-actions-to-trainees-missing-fields-banner)

**Review by ralph-hawkins — APPROVED** [review]

**Pull request merged**